### DISCUSSION TITLE
Station plays first song in list on a continuous loop ignoring the rest in the list.

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#3856] [User Support]
Hello, All.

This just started happening the other evening. The stream plays the first song in the list repeatedly.
It is almost as if it only imports the .ts files for that one song and then loops the heck out of it.
**==UPDATE==**
`reload_mode = 'rounds'` (This will play all songs, but the updates are many songs between.)
`reload_mode = 'seconds'` (This will ONLY play the first track over and over again.)
`reload_mode = 'watch'` (This will ONLY play the first track over and over again, not matter how many times I add another track, it will not reload the list, not even with a 101.m3u file.)
**Another Update**
I read where [Toots](https://github.com/savonet/liquidsoap/issues/492#issuecomment-345894257) said that having better control over your playlist is to have an m3u file.
I tried, but it still replays the first song over and over again.

(One track per line.)
I wanted to try something.
When loading the "101.asp" file in Chrome, it would have all the songs on a single line.
Then I tried using 101.txt and loaded it into Chrome, and each song is on its own line.
But it does not matter; It will still play the first track repeatedly.
**101.txt** (These are 15-second clips to test.)
```
https://demo.example.com/host/Files/Short/01 - Me against the world-001.mp3
https://demo.example.com/host/Files/Short/02 - Lizzy Borden - Shock-001.mp3
https://demo.example.com/host/Files/Short/03 - Lizzy Borden - Outcast-001.mp3
https://demo.example.com/host/Files/Short/04 - Lizzy Borden - Den Of Thieves-001.mp3
https://demo.example.com/host/Files/Short/05 - Lizzy Borden - Visual Lies-001.mp3
https://demo.example.com/host/Files/Short/06 - Lizzy Borden - Eyes Of A Stranger-001.mp3
https://demo.example.com/host/Files/Short/07 - Lizzy Borden - Lord Of The Flies-001.mp3
https://demo.example.com/host/Files/Short/08 - Lizzy Borden - Voyeur (I'm Watching You)-001.mp3
https://demo.example.com/host/Files/Short/09 - Lizzy Borden - Visions-001.mp3
```
If I add a song to the list, say I put it in as the second song, the list will update. After the first song plays, it will play the second song. THEN, it will jump back to the first song and repeatedly play the first song and the new second song.

This is my **101.liq**
```perl
# This is from my original Icecast/Liquidsoap script
# log dir
set ("log.file.path","/var/log/liquidsoap/101.log")
set("scheduler.fast_queues",1)
# Music
music = playlist(mode = 'normal', reload = 1, reload_mode = 'seconds', "https://www.example.com/DJ/101.asp")


# Update title metadata when both artist and title are present.
def hls_meta(m) =
  title = m["title"]
  artist = m["artist"]
  if artist != "" and title != "" then
#    [("title", "#{artist} - #{title}")]
[...list.assoc.remove("artist", m), ("title", "#{artist} - #{title}")]
 else
    m
  end

end

# Start building the feed with music
 radio101 = metadata.map(update=false, hls_meta,music)
 music.on_metadata(print)
#radio101 = music

# ===========================================
# Provide by gAlleb on March 24, 2024
# (https://github.com/savonet/liquidsoap/issues/3825#issuecomment-2018942115)
# ===========================================
# Stream it out
radio101 = mksafe(radio101)

# For Station101
# Define 3 streams with different bitrates. You can make 1 or 2 streams (as you wish).

aac_lofi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="96k",
        profile="aac_low"
    )
)
aac_midfi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="128k",
        profile="aac_low"
    )
)
aac_hifi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="192k",
        profile="aac_low"
    )
)

# Put them all together
hls_streams = [("aac_lofi", aac_lofi), ("aac_midfi", aac_midfi), ("aac_hifi", aac_hifi)]

# Make interesting names for your segments
def hls_segment_name(~position,~extname,stream_name) =
    timestamp = int_of_float(time())
    duration = 4
    "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(playlist="radio101.m3u8",
    segment_duration=4.0,
    segments=5,
    segments_overhead=2,
    segment_name=hls_segment_name,
#
# Please note that you need to indicate here the real paths to your nginx server.
# They would probably be /var/www/html/...
#
#    persist_at="/var/www/html/hls/persist",
#    "/var/www/html/hls/",

#                streams_info=streams_info,
#                persist=true,

    persist_at="/var/www/stream.example.com/hls/101/persist",
               "/var/www/stream.example.com/hls/101/",


    hls_streams,
    radio101
)
``` 
```
#Set permissions on the following folders.
#1st
sudo chown -R www-data:www-data /var/www/stream.example.com
#2nd
sudo usermod -a -G www-data liquid
#3rd
sudo chgrp -R www-data /var/www/stream.example.com/hls
#4th
sudo chmod -R g+rws /var/www/stream.example.com/hls
```

running this.
`ls -alR /var/www/stream.example.com/`

<details><summary>Output</summary>
<p>

/var/www/stream.example.com/:
total 16
drwxr-xr-x 3 www-data www-data 4096 Apr  2 07:41 .
drwxr-xr-x 4 root     root     4096 Apr  2 07:25 ..
drwxrwsr-x 4 www-data www-data 4096 Apr  3 18:13 hls
-rw-r--r-- 1 www-data www-data  362 Apr  2 07:41 index.html

/var/www/stream.example.com/hls:
total 16
drwxrwsr-x 4 www-data www-data 4096 Apr  3 18:13 .
drwxr-xr-x 3 www-data www-data 4096 Apr  2 07:41 ..
drwxrwsr-x 2 www-data www-data 4096 Apr  3 18:26 101
drwxrwsr-x 2 www-data www-data 4096 Apr  3 18:13 102

/var/www/stream.example.com/hls/101:
total 2008
drwxrwsr-x 2 www-data   www-data     4096 Apr  3 18:26 .
drwxrwsr-x 4 www-data   www-data     4096 Apr  3 18:13 ..
-rw-r--r-- 1 liquid liquid 141940 Apr  3 18:26 aac_hifi_4_1712183171_14752.ts
-rw-r--r-- 1 liquid liquid 143068 Apr  3 18:26 aac_hifi_4_1712183175_14753.ts
-rw-r--r-- 1 liquid liquid 142504 Apr  3 18:26 aac_hifi_4_1712183179_14754.ts
-rw-r--r-- 1 liquid liquid 141188 Apr  3 18:26 aac_hifi_4_1712183183_14755.ts
-rw-r--r-- 1 liquid liquid 142692 Apr  3 18:26 aac_hifi_4_1712183187_14756.ts
-rw-r--r-- 1 liquid liquid 141188 Apr  3 18:26 aac_hifi_4_1712183191_14757.ts
-rw-r--r-- 1 liquid liquid    354 Apr  3 18:26 aac_hifi.m3u8
-rw-r--r-- 1 liquid liquid  80464 Apr  3 18:26 aac_lofi_4_1712183171_14752.ts
-rw-r--r-- 1 liquid liquid  80840 Apr  3 18:26 aac_lofi_4_1712183175_14753.ts
-rw-r--r-- 1 liquid liquid  80464 Apr  3 18:26 aac_lofi_4_1712183179_14754.ts
-rw-r--r-- 1 liquid liquid  80464 Apr  3 18:26 aac_lofi_4_1712183183_14755.ts
-rw-r--r-- 1 liquid liquid  80652 Apr  3 18:26 aac_lofi_4_1712183187_14756.ts
-rw-r--r-- 1 liquid liquid  80840 Apr  3 18:26 aac_lofi_4_1712183191_14757.ts
-rw-r--r-- 1 liquid liquid    354 Apr  3 18:26 aac_lofi.m3u8
-rw-r--r-- 1 liquid liquid 110732 Apr  3 18:26 aac_midfi_4_1712183171_14752.ts
-rw-r--r-- 1 liquid liquid 111296 Apr  3 18:26 aac_midfi_4_1712183175_14753.ts
-rw-r--r-- 1 liquid liquid 110356 Apr  3 18:26 aac_midfi_4_1712183179_14754.ts
-rw-r--r-- 1 liquid liquid 111484 Apr  3 18:26 aac_midfi_4_1712183183_14755.ts
-rw-r--r-- 1 liquid liquid 111108 Apr  3 18:26 aac_midfi_4_1712183187_14756.ts
-rw-r--r-- 1 liquid liquid 109416 Apr  3 18:26 aac_midfi_4_1712183191_14757.ts
-rw-r--r-- 1 liquid liquid    359 Apr  3 18:26 aac_midfi.m3u8
-rw-rwSr-- 1 www-data   www-data      238 Apr  3 18:05 radio101.m3u8

/var/www/stream.example.com/hls/102:
total 8
drwxrwsr-x 2 www-data www-data 4096 Apr  3 18:13 .
drwxrwsr-x 4 www-data www-data 4096 Apr  3 18:13 ..
</p>
</details>

<details><summary>101.log</summary>
<p>
2024/04/03 18:33:26 [decoder.ffmpeg:3] Requested content-type for "/tmp/liq-processa9643b.mp3": {audio=pcm(stereo)}

2024/04/03 18:33:26 [decoder.ffmpeg:3] FFmpeg recognizes "/tmp/liq-processa9643b.mp3" as audio: {codec: mp3, 44100Hz, 2 channel(s)}, video: {codec: mjpeg, 600x596, yuvj420p}

2024/04/03 18:33:26 [decoder.ffmpeg:3] Decoded content-type for "/tmp/liq-processa9643b.mp3": {audio=pcm(stereo)}

2024/04/03 18:33:26 [decoder.ffmpeg:3] Requested content-type for "/tmp/liq-processe516d2.mp3": {audio=pcm(stereo)}

2024/04/03 18:33:26 [decoder.ffmpeg:3] FFmpeg recognizes "/tmp/liq-processe516d2.mp3" as audio: {codec: mp3, 44100Hz, 2 channel(s)}, video: {codec: mjpeg, 600x593, yuvj420p}

2024/04/03 18:33:26 [decoder.ffmpeg:3] Decoded content-type for "/tmp/liq-processe516d2.mp3": {audio=pcm(stereo)}

2024/04/03 18:33:27 [101.asp:3] Reloading playlist with URI https://www.example.com/DJ/101.asp.

2024/04/03 18:33:27 [procol.external:3] Failed to find a file extension for "https://www.example.com/DJ/101.asp". Response has unknown mime-type: "text/html" you may want to add it to `settings.http.mime.extnames` and report to us if it is a common one.

</p>
</details> 
The Journal Log shows the same song running in a forever loop.
<details><summary>Journal Log</summary>
<p>
Apr 03 18:33:38 linuxserver liquidsoap[844]: [("album", "W.A.S.P."), ("date", "1984"), ("on_air_timestamp", "1712183618.00"), ("kind", "{audio=pcm(stereo)}"), ("year", "1984"), ("catalog", "at>

Apr 03 18:33:38 linuxserver liquidsoap[844]: [mp3float @ 0x7f8c07a5e700] Could not update timestamps for skipped samples.

Apr 03 18:33:37 linuxserver liquidsoap[844]: [mp3float @ 0x7f8c07a5e200] Could not update timestamps for discarded samples.
</p>
</details> 

Ran a restart
`sudo systemctl restart liquidsoap@101`
<details><summary>Journal Log</summary>
<p>
Apr 03 18:39:00 linuxserver liquidsoap[6419]: [("album", "W.A.S.P."), ("date", "1984"), ("on_air_timestamp", "1712183940.00"), ("kind", "{audio=pcm(stereo)}"), ("year", "1984"), ("catalog", "a>

Apr 03 18:39:00 linuxserver liquidsoap[6419]: [mp3float @ 0x7fbf6d25e700] Could not update timestamps for skipped samples.

Apr 03 18:38:59 linuxserver liquidsoap[6419]: [mpegts @ 0x7fbf91b8aa00] frame size not set

Apr 03 18:38:59 linuxserver liquidsoap[6419]: [mpegts @ 0x7fbf901dda00] frame size not set

Apr 03 18:38:59 linuxserver liquidsoap[6419]: [mpegts @ 0x7fbf920fdc00] frame size not set

</p>
</details>

**Comment by vitoyucepi** [comment]
Hi @carrzkiss,

```
Failed to find a file extension for "https://www.example.com/DJ/101.asp". Response has unknown mime-type: "text/html" you may want to add it to settings.http.mime.extnames and report to us if it is a common one.
```
I can't make this line disappear on my setup, so I think it's normal.

```
playlist(mode = 'normal', reload = 1, reload_mode = 'seconds', "https://www.example.com/DJ/101.asp")
```
This means you'll reload the playlist every second, combined with `mode="normal"` means you'll always play the first item in the list.
If you want to randomize the songs, use `mode="random"` instead, but 1 second is way too often.

> (This will play all songs, but the updates are many songs between.)

I don't get it, but I think once a full playlist is not frequent enough.

> `reload_mode = 'watch'`

For local files only. You can't watch files over http. There is an exception, but I'd like to make this option intentionally hidden.

**Reply by carrzkiss** [reply]
OK, thanks for the `watch`. It will come in handy as I am actually going to be testing a local list.

I do not want to do random; I want to do the list as they are and in order.
How will I be able to do it in order and still reload the list?

The reload it set at 10, I changed that after posting this.

**Reply by vitoyucepi** [reply]
Reload every 10 seconds?
I have to mention that reloading means dropping the rest of the playlist and loading new items.

**Reply by carrzkiss** [reply]
Right now, I am just testing things out and learning how to set this thing up. 
But I am thinking of doing a local load. I have the server set up with a Windows Domain Network Share Mounted right now, and it works. So I can load the music files from that folder.
So, the playlist() will be something along the lines of.
`music = playlist(mode = 'normal', reload = 1, reload_mode = 'seconds', "/home/liquid/liquidsoap/101.m3u")`
The DJ Panel will generate the .m3u file on the website, so I have to set up a Network Share from Linux to the Windows Server next so the files can be created and updated at will.

Using this
`mode = 'normal', reload = 1, reload_mode = 'rounds'`
How often is this reloaded?
Does `reload_mode = 'rounds'` mean that it gets the update after it finishes its list and starts over?

When I first started the list, it had the following in the list.
1st
```
https://demo.example.com/host/Files/David_Lee_Roth/Eat_'Em_and_Smile/09202023150500784.mp3
https://demo.example.com/host/Files/David_Lee_Roth/Skyscraper/09202023175152072.mp3
https://demo.example.com/host/Files/David_Lee_Roth/Skyscraper/09202023175152806.mp3
```
Then, after it started, I added the following:
2nd
```
https://demo.example.com/host/Files/David_Lee_Roth/Eat_'Em_and_Smile/09202023150500784.mp3
https://demo.example.com/host/Files/David_Lee_Roth/Skyscraper/09202023175152072.mp3
https://demo.example.com/host/Files/David_Lee_Roth/Skyscraper/09202023175152806.mp3
https://demo.example.com/host/Files/Def_Leppard/High_'n'_Dry/10072023202224136.mp3
https://demo.example.com/host/Files/Def_Leppard/Hysteria/10082023193116973.mp3
```
Then, a little while later, I changed its order to this.
3rd
```
https://demo.example.com/host/Files/David_Lee_Roth/Eat_'Em_and_Smile/09202023150500784.mp3
https://demo.example.com/host/Files/Def_Leppard/High_'n'_Dry/10072023202224136.mp3
https://demo.example.com/host/Files/David_Lee_Roth/Skyscraper/09202023175152072.mp3
https://demo.example.com/host/Files/Def_Leppard/Hysteria/10082023193116973.mp3
https://demo.example.com/host/Files/David_Lee_Roth/Skyscraper/09202023175152806.mp3
```

The 1st round played entirely, even after the 2nd was added while the first song played.
The 2nd round just finished.
At this moment, the third list is playing, and the second song is shown in the list above.
So, after the last song in the third list plays, it will play the list from the third again. 
That would explain `rounds` (OK, that's a good name for it; it is doing its rounds.)

So, this would be my question.
Let's say I am running a Daily Request. 
I am **three** songs deep into a 20-song list, and someone requests a song before they go into class to get them pumped up for the big exam.
How can I inject that song into the existing list after the current song OR two songs from the currently playing song?
Example"
```
Current Song
==>Inject song here.
```
(OR)
```
Current Song
=>Next Song
==>Inject song here.
```
I know it cannot be done with `reload_mode = 'rounds'`.
What would be the way to go for this type of scenario?

**Reply by carrzkiss** [reply]
> Vito
> Reload every 10 seconds?
> I have to mention that reloading means dropping the rest of the playlist and loading new items.

OK, thank you, did not know that. (learning.)

**Reply by vitoyucepi** [reply]
If you want to inject a song, `request.queue` is the best way to do it.
There's a section about this in [The Book](https://www.liquidsoap.info/book/book.pdf) on page 87.

**Reply by carrzkiss** [reply]
Thanks, Vito.

I will have to figure out how to run a `telnet` command from the website's DJ Panel.
I will likely have to use asp.net to do it.


I am looking at the code. Would it be something like this in my case?
```
music = playlist(mode='normal',reload=1, reload_mode='rounds', "https://www.example.com/DJ/101.m3u")
queue = request.queue()
radio = fallback(track_sensitive=false, [queue, music])
```

would this
`output(radio)`
be this
`output.file.hls(playlist="radio101.m3u8"`

It will be cool to test this out, which I can do through a cmd prompt.

**Reply by vitoyucepi** [reply]
I'd like to point out that you can use the http api instead of telnet.
Check it out at https://www.liquidsoap.info/doc-2.2.4/harbor_http.html.

**Reply by carrzkiss** [reply]
Thank you, Vito.
A lot of options using the `harbor-http`
I like the metadata one.
I just need more information on how to use it.
If I can use it to grab the metadata, I will leave my links the way they are in liq files.
But if it is the same issue as not being able to grab from distant link sources, I will change everything over and work on the local link and file loads, for which I already have the drive mounted.

I will open another discussion on this one.
Thank you.

**Comment by vitoyucepi** [comment]
If you want to collect journal logs without a pager, use the following command
`journalctl --no-pager -u liquidsoap@101`.
If you'd like to view logs without line truncation, pipe it to `less` like this
`journalctl -u liquidsoap@101 | less`, or with the environment variable `SYSTEMD_LESS=FRXMK journalctl -u liquidsoap@101`.

**Reply by carrzkiss** [reply]
I will add them to my list of commands.
love this `journalctl --no-pager -u liquidsoap@101`

**Reply by vitoyucepi** [reply]
I don't always remember the `--no-pager`, so I often pipe something to `cat`, like this `journalctl -u liquidsoap@101 | cat`.

**Reply by carrzkiss** [reply]
OK, I wondered how to do it with a `cat`.
